parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveComparator.java

```diff
@@ -0,0 +1,286 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.schema;
+
+import org.apache.parquet.io.api.Binary;
+import org.junit.Test;
+
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+
+import static org.apache.parquet.schema.PrimitiveComparator.BOOLEAN_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.DOUBLE_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.FLOAT_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.LEXICOGRAPHICAL_BINARY_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.BINARY_AS_SIGNED_INTEGER_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.SIGNED_INT32_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.SIGNED_INT64_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.UNSIGNED_INT32_COMPARATOR;
+import static org.apache.parquet.schema.PrimitiveComparator.UNSIGNED_INT64_COMPARATOR;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class TestPrimitiveComparator {
+
+  @Test
+  public void testBooleanComparator() {
+    boolean[] values = {false, true};
+
+    for (int i = 0; i < values.length; ++i) {
+      for (int j = 0; j < values.length; ++j) {
+        boolean vi = values[i];
+        boolean vj = values[j];
+        int exp = i - j;
+        assertSignumEquals(vi, vj, exp, BOOLEAN_COMPARATOR.compare(vi, vj));
+        assertSignumEquals(vi, vj, exp, BOOLEAN_COMPARATOR.compare(Boolean.valueOf(vi), Boolean.valueOf(vj)));
+      }
+    }
+
+    checkThrowingUnsupportedException(BOOLEAN_COMPARATOR, Boolean.TYPE);
+  }
+
+  @Test
+  public void testSignedInt32Comparator() {
+    testInt32Comparator(SIGNED_INT32_COMPARATOR,
+      Integer.MIN_VALUE,
+      -12345,
+      -1,
+      0,
+      1,
+      12345,
+      Integer.MAX_VALUE);
+  }
+
+  @Test
+  public void testUnsignedInt32Comparator() {
+    testInt32Comparator(UNSIGNED_INT32_COMPARATOR,
+      0,                  // 0x00000000
+      1,                  // 0x00000001
+      12345,              // 0x00003039
+      Integer.MAX_VALUE,  // 0x7FFFFFFF
+      Integer.MIN_VALUE,  // 0x80000000
+      -12345,             // 0xFFFFCFC7
+      -1);                // 0xFFFFFFFF
+  }
+
+  private void testInt32Comparator(PrimitiveComparator<Integer> comparator, int... values) {
+    for (int i = 0; i < values.length; ++i) {
```

Comment:
Also too complicated here. I'd much rather see a bunch of asserts than a function that implements a sort to find out what the comparison should produce.

Reply:
Personally I like the way Gabor implemented this test. By comparing all test values with all others, a lot more cases are covered in an intuitive way than individual asserts would allow without becoming overwhelming. (It does not sort, just compares all values against all other values.)

Reply:
I think it is much too complicated. It's great to compare each value with the others, but it should be straight-forward to read and verify.

I'm -1 on this patch until this is fixed.

Reply:
In the current implementation an average test executes 64 comparisons which I think is required to properly verify. Do you really think I should expand the loops to 64 independent asserts for each test method?
In addition if a specific value behaves strangely and have to be added to the tests it is one more line to add e.g. 8 additional comparisons. I cannot see why you think it would be better to have some separate assertions.

Reply:
Okay, after looking at the implementation I see what's happening here: the values are in ascending order as they are passed to the checks and the validation methods test each combination to make sure. That makes sense and I'm +1 on the method.

The problem is that it needs to be immediately obvious (for reviewers and people updating this later) that this is validating that the comparator's result and that each value is greater than those before it, equal to itself, and less than those after it. When I read this, I saw `testObjectComparator` with a bunch of code in it and immediately thought: way too complicated.

How about naming these validation functions `assertComparatorOrderAscending` or something like that? Also adding a comment about these validation/assertion functions would be really helpful. It's a good idea, it just isn't clear what is happening.